File: heap/3.mergeKSortedLists/js/mergeKSortedLists.js
function ListNode(val, next) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
}

// 堆
const heap = []

/**
 * 入堆
 * @param {ListNode} node
 */
function pushHeap(node) {
    let i = heap.length
    heap.push(node)
    while (i > 0 && heap[i].val < heap[(i - 1) >> 1].val) {
        swap(i, (i - 1) >> 1)
        i = (i - 1) >> 1
    }
}

/**
 * 出堆
 * @return {ListNode}
 */
function popHeap() {
    const res = heap[0]
    heap[0] = heap.at(-1)
    heap.pop()
    let i = 0
    let l = i * 2 + 1
    let best
    while (l < heap.length) {
        if (l + 1 < heap.length && heap[l + 1].val < heap[l].val) {
            best = l + 1
        } else {
            best = l
        }
        if (heap[i].val > heap[best].val) {
            swap(i, best)
        }
        i = best
        l = i * 2 + 1
    }
    return res

}

/**
 * 交换位置
 * @param {number} i
 * @param {number} j
 */
function swap(i, j) {
    let temp = heap[i]
    heap[i] = heap[j]
    heap[j] = temp
}

/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function (lists) {
    if (lists.length === 0) {
        return null
    }
    // 将链表全部入小根堆
    for (let i of lists) {
        if (i) {
            pushHeap(i)
        }
    }
    if (heap.length === 0) {
        return null
    }
    const head = popHeap()
    if (head.next) {
        pushHeap(head.next)
    }
    let node = head
    while (heap.length > 0) {
        let n = popHeap()
        if (n.next) {
            pushHeap(n.next)
        }
        node.next = n
        node = node.next
    }
    return head
};

File: heap/3.mergeKSortedLists/py/mergeKSortedLists.py
from typing import Optional,List
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    # 堆数据区
    data:List[ListNode] = []
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        # 全是空链表
        if not lists:
            return None
        # 链表全入堆"
        for i in lists:
            if i:
                self.pushHeap(i)
        if not self.data:
            return None
        # 记录头节点
        head = self.popHeap()
        if head.next:
            self.pushHeap(head.next)
        print(self.data)
        node = head
        # 取链表节点串起来，直到堆为空
        while self.data:
            n = self.popHeap()
            print(n.val)
            node.next = n
            if n.next:
                self.pushHeap(n.next)
            node = node.next
        return head
            
            
    # 入堆（小根堆)
    def pushHeap(self,node:ListNode):
        i = len(self.data)
        self.data.append(node)
        while i > 0 and self.data[(i-1)>>1].val > self.data[i].val:
            # 比它的父节点小，向上换
            self.data[(i-1)>>1],self.data[i] = self.data[i],self.data[(i-1)>>1]
            i = (i-1)>>1
    # 出堆
    def popHeap(self)->ListNode:
        res = self.data[0]
        self.data[0] = self.data[-1]
        self.data.pop()
        i = 0
        l = i * 2 + 1
        best = None
        while l < len(self.data):
            # 找出两个子节点中更小的
            if l + 1 < len(self.data) and self.data[l].val > self.data[l+1].val:
                best = l + 1
            else:
                best = l
            # 判断它们当中小的是否比当前值还要小
            if self.data[i].val > self.data[best].val:
                # 交换位置
                self.data[i],self.data[best] = self.data[best],self.data[i]
            i = best
            l = i*2 + 1
        return res
            
            
        